core/client/src/main/java/alluxio/client/netty/NettyRemoteBlockReader.java

```diff
@@ -41,51 +45,50 @@
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
 
   private final Bootstrap mClientBootstrap;
-  private final ClientHandler mHandler;
   /** A reference to read response so we can explicitly release the resource after reading. */
   private RPCBlockReadResponse mReadResponse = null;
 
   /**
    * Creates a new {@link NettyRemoteBlockReader}.
-   *
-   * TODO(gene): Creating a new remote block reader may be expensive, so consider a connection pool.
    */
   public NettyRemoteBlockReader() {
-    mHandler = new ClientHandler();
-    mClientBootstrap = NettyClient.createClientBootstrap(mHandler);
+    mClientBootstrap = NettyClient.createClientBootstrap(new ClientHandler());
   }
 
   /**
    * Constructor.
    *
    * @param clientBootstrap bootstrap class of the client channel
-   * @param clientHandler handler of the client channel
    */
-  public NettyRemoteBlockReader(Bootstrap clientBootstrap, ClientHandler clientHandler) {
+  public NettyRemoteBlockReader(Bootstrap clientBootstrap) {
     mClientBootstrap = clientBootstrap;
-    mHandler = clientHandler;
   }
 
   @Override
   public ByteBuffer readRemoteBlock(InetSocketAddress address, long blockId, long offset,
       long length, long lockId, long sessionId) throws IOException {
     SingleResponseListener listener = null;
+    Channel channel = null;
+    Metrics.NETTY_BLOCK_READ_OPS.inc();
     try {
-      ChannelFuture f = mClientBootstrap.connect(address).sync();
-
-      LOG.info("Connected to remote machine {}", address);
-      Channel channel = f.channel();
+      channel = BlockStoreContext.acquireNettyChannel(address, mClientBootstrap);
       listener = new SingleResponseListener();
-      mHandler.addListener(listener);
-      channel.writeAndFlush(new RPCBlockReadRequest(blockId, offset, length, lockId, sessionId));
+      ((ClientHandler) channel.pipeline().last()).addListener(listener);
```

Comment:
It is not sufficient to do channel.pipeline().get(ClientHandler.class) either because a handler can be added to pipeline multiple times. I think a better way is to add a identifier for each handler when adding them. 

But i feel your suggestion is slightly better then last() . So i have changed the code